File: vectors_benchmark/src/engines/vectors.rs
// Copyright (C) 2021 Bosutech XXI S.L.
//
// nucliadb is offered under the AGPL v3.0 and as commercial software.
// For commercial licensing, contact us at info@nuclia.com.
//
// AGPL:
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program. If not, see <http://www.gnu.org/licenses/>.
//

use super::VectorEngine;
use nucliadb_vectors::data_point::{self, DataPointPin, Elem, LabelDictionary, Similarity};
use nucliadb_vectors::data_point_provider::reader::Reader;
use nucliadb_vectors::data_point_provider::writer::Writer;
use nucliadb_vectors::data_point_provider::*;
use nucliadb_vectors::formula::Formula;
use std::time::SystemTime;

lazy_static::lazy_static! {
    static ref FORMULA: Formula = Formula::new();
}

struct Request<'a>(usize, &'a [f32]);
impl<'a> SearchRequest for Request<'a> {
    fn with_duplicates(&self) -> bool {
        true
    }
    fn get_query(&self) -> &[f32] {
        self.1
    }

    fn get_filter(&self) -> &Formula {
        &FORMULA
    }

    fn no_results(&self) -> usize {
        self.0
    }
    fn min_score(&self) -> f32 {
        -1.0
    }
}

impl VectorEngine for Writer {
    fn add_batch(&mut self, batch_id: String, keys: Vec<String>, embeddings: Vec<Vec<f32>>) {
        let temporal_mark = SystemTime::now();
        let similarity = Similarity::Cosine;

        let mut elems = vec![];
        for (key, vector) in keys.into_iter().zip(embeddings.into_iter()) {
            let elem = Elem::new(key, vector, LabelDictionary::new(vec![]), None);
            elems.push(elem);
        }

        let data_point_pin = DataPointPin::create_pin(self.location()).unwrap();
        data_point::create(&data_point_pin, elems, Some(temporal_mark), similarity).unwrap();

        self.add_data_point(data_point_pin).unwrap();
        self.record_delete(batch_id.as_bytes(), temporal_mark);
        self.commit().unwrap();
    }

    fn search(&self, _no_results: usize, _query: &[f32]) {
        unimplemented!()
    }
}

impl VectorEngine for Reader {
    fn add_batch(&mut self, _batch_id: String, _keys: Vec<String>, _embeddings: Vec<Vec<f32>>) {
        unimplemented!()
    }

    fn search(&self, no_results: usize, query: &[f32]) {
        self.search(&Request(no_results, query)).unwrap();
    }
}

File: nucliadb_performance/nucliadb_performance/utils/vectors.py
# Copyright (C) 2021 Bosutech XXI S.L.
#
# nucliadb is offered under the AGPL v3.0 and as commercial software.
# For commercial licensing, contact us at info@nuclia.com.
#
# AGPL:
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as
# published by the Free Software Foundation, either version 3 of the
# License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program. If not, see <http://www.gnu.org/licenses/>.

import httpx
from sentence_transformers import SentenceTransformer  # type: ignore

from .misc import cache_to_disk

ENCODER = None


def get_encoder():
    global ENCODER
    if ENCODER is not None:
        return ENCODER
    ENCODER = SentenceTransformer("all-MiniLM-L6-v2")
    return ENCODER


@cache_to_disk
def compute_vector(sentence: str) -> list[float]:
    encoder = get_encoder()
    return encoder.encode([sentence])[0].tolist()


@cache_to_disk
async def predict_sentence_to_vector(
    predict_url: str, kbid: str, sentence: str
) -> list[float]:
    client = httpx.AsyncClient(base_url=predict_url)
    resp = await client.get(
        "/sentence",
        headers={"X-STF-KBID": kbid},
        params={"text": sentence},
    )
    resp.raise_for_status()
    data = resp.json()
    return data["data"]
